pkg/router/controller/factory/factory.go

```diff
@@ -127,13 +145,40 @@ func (f *RouterControllerFactory) registerInformerEventHandlers(rc *routercontro
 	if f.NamespaceLabels != nil {
 		f.registerSharedInformerEventHandlers(&kapi.Namespace{}, rc.HandleNamespace)
 	}
-	f.registerSharedInformerEventHandlers(&kapi.Endpoints{}, rc.HandleEndpoints)
+	if f.watchEndpoints {
+		f.registerSharedInformerEventHandlers(&kapi.Endpoints{}, rc.HandleEndpoints)
+	} else {
+		f.registerSharedInformerEventHandlers(&discoveryv1beta1.EndpointSlice{}, func(eventType watch.EventType, obj interface{}) {
+			eps := obj.(*discoveryv1beta1.EndpointSlice)
+			if serviceName := endpointSliceServiceName(eps); serviceName == "" {
+				utilruntime.HandleError(fmt.Errorf("EndpointSlice %s/%s has no %q label", eps.Namespace, eps.Name, ServiceNameLabel))
+			} else {
+				objMeta := eps.ObjectMeta.DeepCopy()
+				objMeta.Name = serviceName
+				rc.HandleEndpointSlice(eventType, *objMeta, f.aggregateEndpointSlice(eps.Namespace, serviceName))
+			}
+		})
+	}
 
 	f.registerSharedInformerEventHandlers(&routev1.Route{}, rc.HandleRoute)
 
 	if rc.WatchNodes {
 		f.registerSharedInformerEventHandlers(&kapi.Node{}, rc.HandleNode)
 	}
+
+}
+
+func (f *RouterControllerFactory) aggregateEndpointSlice(namespace, name string) []discoveryv1beta1.EndpointSlice {
+	objType := reflect.TypeOf(&discoveryv1beta1.EndpointSlice{})
+	objs, _ := f.informers[objType].GetIndexer().ByIndex(ServiceNameIndex, path.Join(namespace, name))
+	var fullSet []discoveryv1beta1.EndpointSlice
+
+	for i := range objs {
+		eps := objs[i].(*discoveryv1beta1.EndpointSlice)
+		fullSet = append(fullSet, *eps.DeepCopy())
```

Comment:
Is using a copy necessary here?

Reply:
It comes out of the store. But will look to see if we mutate at all and if not we can use it as-is.